snappiershot/snapshot/_raises.py

```diff
@@ -0,0 +1,78 @@
+""" Context class used to track and snapshot raised exceptions.
+
+Mimics and is inspired by `pytest.raises`.
+"""
+from types import TracebackType
+from typing import (
+    TYPE_CHECKING,
+    Generic,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+if TYPE_CHECKING:  # pragma: no cover
+    # Used to avoid circular imports.
+    from .snapshot import Snapshot
+
+_E = TypeVar("_E", bound=BaseException)
+_ExceptionTypes = Union[Type[_E], Tuple[Type[_E], ...]]
+
+
+class _RaisesContext(Generic[_E]):
+    """ Context class used to track and snapshot raised exceptions. """
+
+    def __init__(
+        self, snapshot: "Snapshot", expected_exceptions: _ExceptionTypes, update: bool
+    ):
+        """
+        Args:
+            snapshot: The snapshot object. Used to perform the actual snapshotting of the
+              raised exception. This object must have the metadata preset.
+            expected_exceptions: A tuple of exception types that are expected to be raised.
+            update: Whether to update the snapshot or not.
+        """
+        self.snapshot = snapshot
+        self.expected_exceptions = expected_exceptions
+        self.update = update
+
+        # Tracks whether the assertion was matched. Used for testing.
+        self._assert_match = False
+
+    def __enter__(self) -> "_RaisesContext":
+        """ Enter the context. """
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> bool:
+        """ Exit the context, checking exception types (if raised) and snapshotting.
+
+        Args:
+            exc_type: The type of exception raised within the context (if exists).
+            exc_val: The exception raised within the context (if exists).
+            exc_tb: The traceback of the exception raised with the context (if exists).
+
+        Raises:
+            ValueError: If no exception was raised.
+            TypeError: If an unexpected type was raised.
+            AssertionError: If the exception does not match the snapshot.
+        """
+        if exc_type is None:
+            raise ValueError("No exception raised. ")
+        if not issubclass(exc_type, self.expected_exceptions):
```

Comment:
do we want this `if` clause? I would think not. If the exception type doesn't match the snapshot, then the snapshot will fail. I do not think an additional error is the appropriate behaviour here. If any error is caught, it should be compared with the snapshot, no?

Reply:
two reasons:
1) This is what `pytest` does, and if we are mimicking the `pytest.raises` call we should mimic that as well.
2) This would preemptively check that the correct error is being snapshotted when the snapshot is written/updated.